src/scripts/clipperUI/components/modeButtonSelector.tsx

```diff
@@ -31,101 +26,142 @@ class ModeButtonSelectorClass extends ComponentBase<{}, ClipperStateProp> {
-			myMode={ClipMode.Selection} tabIndex={43} selected={currentMode === ClipMode.Selection}
-			onModeSelected={this.onModeSelected.bind(this) }
-			tooltipText={Localization.getLocalizedString("WebClipper.ClipType.Selection.Button.Tooltip")}/>;
+		return {
+			imgSrc: ExtensionUtils.getImageResourceUrl("select.png"),
+			label: Localization.getLocalizedString("WebClipper.ClipType.Selection.Button"),
+			myMode: ClipMode.Selection,
+			selected: currentMode === ClipMode.Selection,
+			onModeSelected: this.onModeSelected.bind(this),
+			tooltipText: Localization.getLocalizedString("WebClipper.ClipType.Selection.Button.Tooltip")
+		};
 	}
 
-	private getBookmarkModeButton(currentMode: ClipMode) {
-		if (this.props.clipperState.pageInfo.rawUrl.indexOf("file:///") === 0) {
+	private getBookmarkButtonProps(currentMode: ClipMode): PropsForModeElementNoAriaGrouping {
+		if (this.props.clipperState.pageInfo.rawUrl.slice(0, 8) === "file:///") {
 			return undefined;
 		}
-		return <ModeButton imgSrc={ExtensionUtils.getImageResourceUrl("bookmark.png") }
-			label={Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button") }
-			myMode={ClipMode.Bookmark} tabIndex={44} selected={currentMode === ClipMode.Bookmark}
-			onModeSelected={this.onModeSelected.bind(this) }
-			tooltipText={Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button.Tooltip") } />;
+
+		return {
+			imgSrc: ExtensionUtils.getImageResourceUrl("bookmark.png"),
+			label: Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button"),
+			myMode: ClipMode.Bookmark,
+			selected: currentMode === ClipMode.Bookmark,
+			onModeSelected: this.onModeSelected.bind(this),
+			tooltipText: Localization.getLocalizedString("WebClipper.ClipType.Bookmark.Button.Tooltip")
+		};
 	}
 
-	public render() {
+	private getListOfButtons() {
 		let currentMode = this.props.clipperState.currentMode.get();
 
+		let buttonProps = [
+			this.getFullPageButtonProps(currentMode),
+			this.getRegionButtonProps(currentMode),
+			this.getAugmentationButtonProps(currentMode),
+			this.getSelectionButtonProps(currentMode),
+			this.getBookmarkButtonProps(currentMode),
+			this.getPdfButtonProps(currentMode),
+		];
+
+		let visibleButtons = [
+			this.getScreenReaderOnlyElementPropsThatAnnouncesCurrentMode(currentMode),
+		];
+
+		if (buttonProps) {
+			let propsForVisibleButtons = buttonProps.filter(attributes => !!attributes);
+			for (let i = 0; i < propsForVisibleButtons.length; i++) {
+				let attributes = propsForVisibleButtons[i];
+				let ariaPos = i + 1;
+				visibleButtons.push(<ModeButton imgSrc={attributes.imgSrc} label={attributes.label} myMode={attributes.myMode}
+								selected={attributes.selected} onModeSelected={attributes.onModeSelected}
+								tooltipText={attributes.tooltipText} aria-setsize={propsForVisibleButtons.length}
+								aria-posinset={ariaPos} tabIndex={attributes.selected ? 40 : undefined} />);
+			}
+			return visibleButtons;
+		}
```

Comment:
It's possible that this method does not return a value

Reply:
(see above comment to add return type to the method, that will add stricter compile time checking which would prevent this issue)